### ISSUE TITLE
COPY cache not invalidating correctly with recursive copy

### ISSUE THREAD

**Issue by samuelcolvin** [docker/docker#17906]
my dockerfile has 

```dockerfile
...
COPY code/ /src/code/
...
```

where `code` is a directory containing code files (including sub directories).

Making changes to these files is **not** invalidating the cache, and `docker build` is showing:

```
Step 11 : COPY code/ /src/code/
 ---> Using cache
 ---> 638bbc8d86ff
```

This problem is contrary to all the other issues in this area which complain invalidation is too agressive. IMHO this is much more serious as the resultant image is objectively wrong: it contains old versions of the code.

I've tried both `COPY` and `ADD` with the same problem, I'm on `Docker version 1.9.0, build 76d6bc9`, Ubuntu 15.10.

**Comment by samuelcolvin** [comment]
more details system information as requested:

```
$ || docker version
Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:54:53 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:54:53 UTC 2015
 OS/Arch:      linux/amd64
$ || docker info
Containers: 10
Images: 258
Server Version: 1.9.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 278
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.0-18-generic
Operating System: Ubuntu 15.10
CPUs: 8
Total Memory: 15.63 GiB
Name: XXX
ID: GYIK:CCVB:LLSG:WZ2D:G5MG:A7WA:RVN3:SYEF:F2N4:UZAH:476G:BZPC
WARNING: No swap limit support
$ || uname -a
Linux X 4.2.0-18-generic #22-Ubuntu SMP Fri Nov 6 18:25:50 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by thaJeztah** [comment]
Could this be the same issue as https://github.com/docker/docker/issues/17290 (which was resolved by https://github.com/docker/docker/pull/17710) if it is, this will be fixed in 1.9.1

**Comment by samuelcolvin** [comment]
could be related but there are no symlinks anywhere in my project.

**Comment by thaJeztah** [comment]
ping @tonistiigi could you have a look?

**Comment by carlossg** [comment]
This is a regression in 1.9.0, 1.8.3 works fine

# 1.8.3

```
$ cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1 >  deps/dna/dna/util.py
$ /usr/local/Cellar/docker/1.8.3/bin/docker version
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.5.1
 Git commit:   f4bf5c7
 Built:        Thu Oct 15 09:34:18 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   f4bf5c7
 Built:        Mon Oct 12 18:01:15 UTC 2015
 OS/Arch:      linux/amd64
$ /usr/local/Cellar/docker/1.8.3/bin/docker build .
Sending build context to Docker daemon 7.865 MB
Step 0 : FROM debian
 ---> 91bac885982d
...
Step 8 : COPY deps/dna/scripts /usr/local/bin/
 ---> Using cache
 ---> 00603962a3ed
Step 9 : COPY deps/dna/boto $PYTHON_DIR/boto
 ---> Using cache
 ---> ca8d6ec59c5e
Step 10 : COPY deps/dna/django $PYTHON_DIR/django
 ---> Using cache
 ---> b322ca396597
Step 11 : COPY deps/dna/dna $PYTHON_DIR/dna
 ---> ccea3794d7a5
Removing intermediate container 48ed471c5fed
Step 12 : COPY deps/something/scripts /usr/local/bin/
 ---> ba76e1784f18
Removing intermediate container 61b640046157
Step 13 : COPY deps/something/argcomplete $PYTHON_DIR/argcomplete
 ---> 72ba7b855022
Removing intermediate container 23dfae81b0b2
Step 14 : COPY deps/something/jce $PYTHON_DIR/jce
 ---> 684776e00b92
Removing intermediate container d3a2cbf46257
Step 15 : COPY deps/something-templates /usr/local/share/something/setup-templates
 ---> bbcb82978943
Removing intermediate container 1d9cba627118
Step 16 : COPY deps/something-tests /usr/local/share/something/tests
 ---> 26ca4c16a775
Removing intermediate container 78aaea53afe7
Step 17 : COPY deps/something/VERSIONS /usr/local/share/something/VERSIONS
 ---> 9209919ac844
Removing intermediate container 7c4590265901
Step 18 : WORKDIR /var/something
 ---> Running in 866570a48917
 ---> aca36196b118
Removing intermediate container 866570a48917
Step 19 : ENTRYPOINT /usr/local/bin/something
 ---> Running in 065fbb08f7cc
 ---> e7d2d1eaf0b0
Removing intermediate container 065fbb08f7cc
Successfully built e7d2d1eaf0b0
```

# 1.9.0

```
$ cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1 >  deps/dna/dna/util.py
$ docker version
Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.5.1
 Git commit:   76d6bc9
 Built:        Tue Nov  3 21:18:00 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   76d6bc9
 Built:        Tue Nov  3 19:20:09 UTC 2015
 OS/Arch:      linux/amd64
$ docker build .
Sending build context to Docker daemon 7.865 MB
Step 1 : FROM debian
 ---> 91bac885982d
...
Step 9 : COPY deps/dna/scripts /usr/local/bin/
 ---> Using cache
 ---> d3cf4d92d2cc
Step 10 : COPY deps/dna/boto $PYTHON_DIR/boto
 ---> Using cache
 ---> 65fed087a369
Step 11 : COPY deps/dna/django $PYTHON_DIR/django
 ---> Using cache
 ---> 970f3420d75f
Step 12 : COPY deps/dna/dna $PYTHON_DIR/dna
 ---> Using cache
 ---> 9d1c58f0b5a1
Step 13 : COPY deps/something/scripts /usr/local/bin/
 ---> Using cache
 ---> f35167d5abea
Step 14 : COPY deps/something/argcomplete $PYTHON_DIR/argcomplete
 ---> Using cache
 ---> 6a3bc642f467
Step 15 : COPY deps/something/jce $PYTHON_DIR/jce
 ---> Using cache
 ---> 31f4f66a3edf
Step 16 : COPY deps/something-templates /usr/local/share/something/setup-templates
 ---> Using cache
 ---> 531358f8aaac
Step 17 : COPY deps/something-tests /usr/local/share/something/tests
 ---> Using cache
 ---> 8f078d0ce4f7
Step 18 : COPY deps/something/VERSIONS /usr/local/share/something/VERSIONS
 ---> Using cache
 ---> 27636396d336
Step 19 : WORKDIR /var/something
 ---> Using cache
 ---> b9b76ebc27a5
Step 20 : ENTRYPOINT /usr/local/bin/something
 ---> Using cache
 ---> bf3faa309566
Successfully built bf3faa309566
```

**Comment by tiborvass** [comment]
@samuelcolvin @carlossg is there a small reproducible case we could use to test?

**Comment by samuelcolvin** [comment]
Not sure this helps much, I've created a small example that should demonstrate the problem:

https://github.com/samuelcolvin/docker-copy-bug

However it's working correctly and copying the directory each time.

I've tried everything I can think of to reproduce the bug but with no luck.

I thought it might be something to do with starting building the image using 1.6 then updating to 1.9, but I've tried switching between the two version without being able to reproduce the issue.

Let me know if there's any more details you need about the original project where the problem occurred? I've tried duplicating as much of it as I can in the test but to no avail.

**Comment by samuelcolvin** [comment]
Ok, at last I've managed to replicate the issue. Two things had to be changed for it to occur:

1. the `env` directory had to exist next to `Dockerfile` it didn't have an effect if it was empty, i didn't check what needed to be in it, i just ran `virtualenv env`
2. the directory being copied needed to be called `gen` this seems odd and makes me think the problem is that somehow docker's cache on my machine has got messed up, so `gen` (the original name of the code directory in the first project where the problem occurred) is significant.

Output of `./test.sh`:

```
cat gen/file.py: date = '18:56:37.756251477'
Sending build context to Docker daemon 23.63 MB
Step 1 : FROM ubuntu:15.10
 ---> 337d0fd75c93
Step 2 : COPY gen/ /src/gen/
 ---> Using cache
 ---> 1f87d344deea
Successfully built 1f87d344deea

-------------------------
updating gen/file.py
-------------------------

cat gen/file.py: date = '18:56:38.642704571'
Sending build context to Docker daemon 23.63 MB
Step 1 : FROM ubuntu:15.10
 ---> 337d0fd75c93
Step 2 : COPY gen/ /src/gen/
 ---> Using cache
 ---> 1f87d344deea
Successfully built 1f87d344deea
```

**Comment by carlossg** [comment]
ok, I realized that I had a broken symlink in a subdir of my workdir, not the subdir being copied though, another one.
So it might be #17290

**Comment by cpuguy83** [comment]
Yep I think it is. @carlossg Any chance you can try from master? https://master.dockerproject.org

**Comment by tiborvass** [comment]
I tested this with 1.9.1-rc1 and it is fixed by #17710
